### PR TITLE
Add a Zombied event to clients

### PR DESCRIPTION
**Pull request by alexhorner**
# Summary
Adds a Zombied event to the DiscordClient and DiscordShardedClient for heartbeat failures

# Details
When the heartbeat failure zombie message is triggered on the log, no event was present to execute a response. The Zombied event has now been added to allow the capture of this occurence.

# Changes proposed
* Add an event for when enough heartbeat failures have occured to consider the connection zombied

# Notes
I have created this event as over the last few days I have seena  lot of zombied connections on my bot. I now have this running in prod. It has not naturally triggered yet, so I cannot confirm its functionality, but if there is a good way to test this, please let me know. If the event occurs naturally I shall reply. Hopefully that is acceptable for the contents of this PR

### CODE REVIEW

**Comment by Naamloos** [comment]
I'm not 100% sure if my memory serves me correctly, but I think I remember that someone unplugged their LAN to test reconnects.. maybe the same works here? 😜

**Comment by alexhorner** [comment]
> 
> 
> I'm not 100% sure if my memory serves me correctly, but I think I remember that someone unplugged their LAN to test reconnects.. maybe the same works here? 😜

I'll shove a test copy of my bot on a laptop and see what gives!

**Comment by alexhorner** [comment]
Seems like disconnecting doesn't cause heartbeat failures.

As discussed in discord, attempting event filtering proxy

File: DSharpPlus/Clients/DiscordShardedClient.cs [hunk 1]
```diff
@@ -540,6 +541,7 @@ private void HookEventHandlers(DiscordClient client)
             client.MessageReactionRemovedEmoji += this.Client_MessageReactionRemovedEmoji;
             client.WebhooksUpdated += this.Client_WebhooksUpdate;
             client.Heartbeated += this.Client_HeartBeated;
+            client.Zombied += this.Client_Zombied;
```
Comment (Lulalaby) [inline DSharpPlus/Clients/DiscordShardedClient.cs:544]: Add this to UnhookEventHandlers too

Comment (alexhorner) [reply DSharpPlus/Clients/DiscordShardedClient.cs:544]: Will do this

**Comment by alexhorner** [comment]
Just had my first instance of it live, working fine. Gonna switch to draft and fix the above review then request review again

File: DSharpPlus/Clients/DiscordShardedClient.cs (hunk 1 above)
Comment (alexhorner) [reply DSharpPlus/Clients/DiscordShardedClient.cs:551]: This has now been done in the latest commit

File: DSharpPlus/Clients/DiscordClient.WebSocket.cs [hunk 2]
```diff
@@ -404,14 +404,32 @@ internal async Task SendHeartbeatAsync(long seq)
...
             var more_than_5 = Volatile.Read(ref this._skippedHeartbeats) > 5;
             var guilds_comp = Volatile.Read(ref this._guildDownloadCompleted);
+
             if (guilds_comp && more_than_5)
             {
                 this.Logger.LogCritical(LoggerEvents.HeartbeatFailure, "Server failed to acknowledge more than 5 heartbeats - connection is zombie");
+
+                var args = new ZombiedEventArgs
+                {
+                    Failures = Volatile.Read(ref this._skippedHeartbeats),
```
Comment (IDoEverything) [inline DSharpPlus/Clients/DiscordClient.WebSocket.cs:414]: Does this need to be `Volatile.Read`?

Comment (alexhorner) [reply DSharpPlus/Clients/DiscordClient.WebSocket.cs:414]: > 
> 
> Does this need to be `Volatile.Read`?

I only made it so as the other reads are also volatile. If you believe it should not be I can change that

Comment (IDoEverything) [reply DSharpPlus/Clients/DiscordClient.WebSocket.cs:414]: If for the sake of consistency, sure

**Review by IDoEverything — APPROVED** [review]

**Pull request merged**